explorer/interpreter/type_checker.cpp

```diff
@@ -392,6 +392,56 @@ static auto IsConcreteType(Nonnull<const Value*> value) -> bool {
   return IsType(value) && !TypeContainsAuto(value);
 }
 
+// Returns whether the given value is template-dependent, that is, if it
+// depends on any template paramaeter.
+static auto IsTemplateDependent(Nonnull<const Value*> value) -> bool {
+  // A VariableType is template dependent if it names a template binding.
+  if (auto* var_type = dyn_cast<VariableType>(value)) {
+    return var_type->binding().binding_kind() ==
+           GenericBinding::BindingKind::Template;
+  }
+
+  static constexpr auto is_dependent_value = [](auto&& x) {
+    if constexpr (std::is_convertible_v<decltype(x), const Value*>) {
+      return IsTemplateDependent(x);
+    }
+    return false;
+  };
+
+  // Any other value is template dependent if any part of it is.
+  return value->Visit<bool>([](auto* derived_value) {
```

Comment:
```suggestion
  return value->Visit([](auto* derived_value) -> bool {
```

Would indicating the return type of the lambda allow dropping the `<bool>`? I think indicating the return type this way may be a little clearer.

Reply:
No, we can't drop the `<bool>` currently -- `Visit` doesn't infer the return type from the callable. We could probably fix that by asking what type the callable returns for some arbitrary `Value`-derived class and assuming it'll be the same for all derived classes. I'll look into that separately.